src/osd/PG.cc

```diff
@@ -2639,115 +2642,89 @@ void PG::_update_calc_stats()
-    // degraded and account as misplaced.
-    for (const auto& peer : peer_info) {
-      if (actingbackfill.find(peer.first) == actingbackfill.end()) {
-	object_copies += peer.second.stats.stats.sum.num_objects;
-	misplaced += peer.second.stats.stats.sum.num_objects;
-	++num_misplaced;
-      }
+    // Objects missing from up nodes, sorted by # objects.
+    set<pair<int64_t,pg_shard_t>> missing_target_objects;
+    // Objects missing from nodes not in up, sort by # objects
+    set<pair<int64_t,pg_shard_t>> acting_source_objects;
+    // Objects missing on any other peers
+    set<pair<int64_t,pg_shard_t>> peer_source_objects;
+
+    int64_t missing;
+
+    // Primary first
+    missing = pg_log.get_missing().num_missing();
+    info.stats.stats.sum.num_objects_missing_on_primary = missing;
+    assert(actingbackfill.count(pg_whoami));
+    if (upset.count(pg_whoami)) {
+      missing_target_objects.insert(make_pair(missing, pg_whoami));
+    } else {
+      acting_source_objects.insert(make_pair(missing, pg_whoami));
     }
-    if (object_copies)
-      dout(20) << __func__ << " objects not part of up/acting " << object_copies << dendl;
 
-    // Objects backfilled, sorted by # objects.
-    set<pair<int64_t,pg_shard_t>> backfill_target_objects;
-
-    assert(!actingbackfill.empty());
-    for (set<pg_shard_t>::iterator i = actingbackfill.begin();
-	 i != actingbackfill.end();
-	 ++i) {
-      const pg_shard_t &p = *i;
-      bool in_up = (upset.find(p) != upset.end());
-
-      // recovery       Compute total objects excluding num_missing
-      //   - not in up  Compute misplaced objects excluding num_missing
-      // backfill       Compute total objects already backfilled
-      if (!is_backfill_targets(p)) {
-        unsigned osd_missing, osd_objects;
-        // primary handling
-        if (p == pg_whoami) {
-          osd_missing = pg_log.get_missing().num_missing();
-          info.stats.stats.sum.num_objects_missing_on_primary =
-              osd_missing;
-        } else {
-          assert(peer_missing.count(p));
-          osd_missing = peer_missing[p].num_missing();
+    // All other peers
+    for (auto& peer : peer_info) {
+      missing = 0;
+      // Backfill targets always track num_objects accurately
+      // all other peers track missing accurately.
+      if (is_backfill_targets(peer.first)) {
+	missing = num_objects - peer.second.stats.stats.sum.num_objects;
+        if (missing < 0) missing = 0;
+      } else {
+	if (peer_missing.count(peer.first)) {
+	  missing = peer_missing[peer.first].num_missing();
+	} else {
+	  dout(20) << __func__ << " no peer_mssing found for " << peer.first << dendl;
         }
-
-        osd_objects = MAX(0, num_objects - osd_missing);
-        object_copies += osd_objects;
-        // Count non-missing objects not in up as misplaced
-	if (!in_up) {
-	  misplaced += osd_objects;
-	  ++num_misplaced;
-	}
+      }
+      peer.second.stats.stats.sum.num_objects_missing = missing;
```

Comment:
not sure we should be in the business of changing the peer stats here?  i worry we'll accidentally rely on it elsewhere

Reply:
This function has always had done this.